### PR TITLE
Add Joiner component (Transformer part 2)

### PR DESCRIPTION
**Pull request by hacodeorg**
3rd PR in the new survey pipeline series. 

[Task PLC-47](https://codedotorg.atlassian.net/browse/PLC-47) -> [1st PR](https://github.com/code-dot-org/code-dot-org/pull/28736) -> [2nd PR](https://github.com/code-dot-org/code-dot-org/pull/28766) -> [3rd PR](https://github.com/code-dot-org/code-dot-org/pull/28770) -> [4th PR](https://github.com/code-dot-org/code-dot-org/pull/28816) -> [5th PR](https://github.com/code-dot-org/code-dot-org/pull/28823) -> [6th](https://github.com/code-dot-org/code-dot-org/pull/28846)

This code takes the outputs from the Parser ([2nd PR](https://github.com/code-dot-org/code-dot-org/pull/28766)), joins answers to questions and flatten the results so they can be summarized later.

- Input
  - Survey input (questions): `Hash{form_id => {question_id => question_content}}` (Question id 4 is a matrix question.)
    ```
    {
      1 => {
        '4' => {
          type: 'matrix', name: 'multiQuestions', text: 'Answer all sub questions',
          sub_questions: ['Sub question 1', 'Sub question 2', 'Sub question 3'],
          options: ['Option 1', 'Option 2', 'Option 3'],
          option_map: {'Option 1' => 1, 'Option 2' => 2, 'Option 3' => 3},
          answer_type: 'multiSelect'
        }
      }
    }
    ```
  - Submission input (answers): `Hash{form_id => {submission_id => submission_content}}`
    ```
    {
      1 => {
        91 => {
          workshop_id: 9999,
          user_id: 1,
          day: 0,
          session_id: nil,
          facilitator_id: nil,
          answers: {
              '4' => [
                {text: 'Sub question 1', answer: 'Option 1'},
                {text: 'Sub question 2', answer: 'Option 2'},
                {text: 'Sub question 3', answer: 'Option 3'}
              ]
          }
        }
      }
    }
    ```

- Output `[Array<Hash>]`. The hash has keys from submission, question and answer.
    ```
    [{:workshop_id=>9999, :user_id=>1, :day=>0, :session_id=>nil, :facilitator_id=>nil, 
      :form_id=>1, :submission_id=>91, 
      :type=>"radio", :name=>"multiQuestions_3e5220bb4661a8e", 
      :text=>"Answer all sub questions->Sub question 1", 
      :options=>["Option 1", "Option 2", "Option 3"], 
      :option_map=>{"Option 1"=>1, "Option 2"=>2, "Option 3"=>3}, 
      :answer_type=>"singleSelect", :max_value=>3, 
      :parent=>"multiQuestions", 
      :qid=>"Q4_3e5220bb4661a8e", 
      :answer=>"Option 1"}, 
    {:workshop_id=>9999, :user_id=>1, :day=>0, :session_id=>nil, :facilitator_id=>nil, 
      :form_id=>1, :submission_id=>91, 
      :type=>"radio", :name=>"multiQuestions_eda0b8cecd575ef", 
      :text=>"Answer all sub questions->Sub question 2", 
      :options=>["Option 1", "Option 2", "Option 3"], 
      :option_map=>{"Option 1"=>1, "Option 2"=>2, "Option 3"=>3}, 
      :answer_type=>"singleSelect", :max_value=>3, 
      :parent=>"multiQuestions", 
      :qid=>"Q4_eda0b8cecd575ef", 
      :answer=>"Option 2"}, 
    {:workshop_id=>9999, :user_id=>1, :day=>0, :session_id=>nil, :facilitator_id=>nil, 
      :form_id=>1, :submission_id=>91, 
      :type=>"radio", :name=>"multiQuestions_99cddc94b8c1d6e", 
      :text=>"Answer all sub questions->Sub question 3", 
      :options=>["Option 1", "Option 2", "Option 3"], 
      :option_map=>{"Option 1"=>1, "Option 2"=>2, "Option 3"=>3}, 
      :answer_type=>"singleSelect", :max_value=>3, 
      :parent=>"multiQuestions", 
      :qid=>"Q4_99cddc94b8c1d6e", 
      :answer=>"Option 3"}]
    ```

How tested
- `bundle exec rails test test/lib/pd/survey_pipeline/daily_survey_joiner_test.rb`

### CODE REVIEW

File: dashboard/lib/pd/survey_pipeline/daily_survey_joiner.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    # Compute a descendant key based on original value and sub_value digest
+    def self.compute_descendant_key(value, sub_value)
+      "#{value}_#{compute_str_digest(sub_value)}"
+    end
+
+    # Compute message digest using SHA256 and cut it to a fixed length.
+    def self.compute_str_digest(str)
+      # SHA256 returns a 256-bit digest, which is 64-character string in hex
+      Digest::SHA256.hexdigest(str || '')[0, STRING_DIGEST_LENGTH - 1]
```
Comment (islemaster) [inline dashboard/lib/pd/survey_pipeline/daily_survey_joiner.rb]: Please consider switching from SHA-256 to a non-cryptographic hash like [xxHash](http://www.xxhash.com/) for this use case.  There's no need for this hash to be secure, we only need to minimize collisions - and xxHash is almost 20x faster than SHA-256, which may become relevant once we're working on very large sets of survey data.

We [already use xxHash](https://github.com/code-dot-org/code-dot-org/blob/8c2a2e6ab57e5a23604ea599ab2ce458b4e9b34d/shared/middleware/helpers/sharded_redis_factory.rb#L45) ([Gemfile](https://github.com/code-dot-org/code-dot-org/blob/ab039b366418d821c41b2cd2d9c948b230e336d7/Gemfile#L48)) so you won't be adding a dependency. (You shouldn't need JumpHash in this case.)

File: dashboard/test/lib/pd/survey_pipeline/daily_survey_joiner_test.rb [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            }
+          )
+        }
+      }
+
+      @submissions_context.merge!(form_id: form_id, submission_id: submission_id)
+    end
+
+    test 'can join submission and question' do
+      # Expected flatten results for non-matrix questions
```
Comment (islemaster) [inline dashboard/test/lib/pd/survey_pipeline/daily_survey_joiner_test.rb]: You've got a large `setup_all` for one test case here!

I'd like to see the matrix question behavior in a separate test case from other question types.

**Review by islemaster — APPROVED** [review]
Looks good, please consider comments for improvement.

File: dashboard/lib/pd/survey_pipeline/daily_survey_joiner.rb (hunk 1 above)
Comment (hacodeorg) [reply dashboard/lib/pd/survey_pipeline/daily_survey_joiner.rb]: oh great. I'll take a look at xxHash

File: dashboard/test/lib/pd/survey_pipeline/daily_survey_joiner_test.rb (hunk 2 above)
Comment (hacodeorg) [reply dashboard/test/lib/pd/survey_pipeline/daily_survey_joiner_test.rb]: The expectation is more tests will be added later. I can move the setup into the test and later split it out if it's going to be shared among tests.

Comment (hacodeorg) [reply dashboard/test/lib/pd/survey_pipeline/daily_survey_joiner_test.rb]: Split into 2 tests, 1 for non-matrix and 1 for matrix question.

**Pull request merged**